### ISSUE TITLE
[suggestion] System package checks in tools.system.Zypper should be capability-aware

### ISSUE THREAD

**Issue by neobrain** [conan-io/conan#15335]
### What is your suggestion?

tl;dr: Conan should add `--whatprovides` to the `check_command` used for Zypper, so that renamed packages can be detected.

Distributions may rename/merge system packages from time. To make sure old install instructions still work, a system called *capabilities* is used to allow the package manager to automatically install the right thing. For example, openSUSE doesn't have a `glproto-devel` package, but instead there's an `xorgproto-devel` package with the `glproto-devel` *capability*. That way, the user can enter `zypper install glproto-devel` to install `xorgproto-devel`. Conan currently is *not* capability-aware, hence system packages always must be referred to by their actual package name.

Conan Center is hitting this problem in `opengl/system`, which has a system package requirement to `glproto-devel`. That requirement is always detected as "not installed" on openSUSE, even if the user manually installs `xorgproto-devel` (see https://github.com/conan-io/conan-center-index/issues/18912). A workaround is to set `tools.system.package_manager:mode=install` and `tools.system.package_manager:sudo=True`, which will still fail the check but will then invoke a no-op `zypper install`.

This is because Conan's `check_command` (conan/tools/system/package_manager.py) uses `rpm -q <package_name>` under the hood, which strictly searches for package names. Luckily, rpm provides a flag to search for capabilities instead: ``rpm -q --whatprovides <package_name>``

### Have you read the CONTRIBUTING guide?

- [x] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @neobrain 

Thanks for the suggestion.
I think it makes sense and it could be added to Conan, a couple of quick questions:
- How long has been ``--whatprovides`` been around? If it is a very new thing, then it can have the opposite effect and break existing users with older versions of zypper
- Is there any other risk that adding this flag could have?

If you are using that distro, and if you are willing to add this yourself with a PR, that would be great, because you might be able to test in a real system that this works, our CI can only check some mocked things, but not that it really works. If not, don't worry, we will try to figure it out. Thanks!

**Comment by neobrain** [comment]
Thanks for the quick reply!

I don't know if there's any risks associated with this flag, i.e. in particularly I haven't done any deep dive to verify this provides strictly a superset of the normal `-q` mode.

I did some brief archeology on the flag's history though: http://ftp.rpm.org/api/4.4.2.2/dependencies.html mentions it, and rpm 4.4.22 was released back in 2007. I didn't check how long before that it was added in the first place, since that already seems far enough in the past :)

> if you are willing to add this yourself with a PR

I wouldn't mind, but I haven't signed the CLA and can't justify the time commitment for a legal review considering the scope of this change. Happy to do testing though!

**Comment by neobrain** [comment]
For what it's worth, I've been running Conan with my proposed change over the past months and haven't encountered any issues.

**Comment by memsharded** [comment]
This would be basically changing this?
```python
class Zypper(_SystemPackageManagerTool):
    tool_name = "zypper"
    install_command = "{sudo}{tool} --non-interactive in {packages}"
    update_command = "{sudo}{tool} --non-interactive ref"
    check_command = "rpm -q --whatprovides {package}"
```

Even if this seems convenient, I am a bit concerned about the different behavior. It seems the other wrappers for other system package managers are looking for exact package matches, not for alternative providers or other changed names when system package manager changes. 
I am not saying it is not possible to change it, just trying to understand the whole picture, maybe @czoido recall something?

**Comment by czoido** [comment]
Hi @neobrain,

Thanks a lot for reporting. I think this could be solved using the [install_substitutes](https://docs.conan.io/2/reference/tools/system/package_manager.html#conan.tools.system.package_manager.Zypper.install_substitutes) method from the Conan system package manager. It would just be changing this in the opengl recipe: https://github.com/conan-io/conan-center-index/blob/0bff285ebad4febf9e31a1efbf294282a7bdc122/recipes/opengl/all/conanfile.py#L39

to something like this:

```
zypper.install_substitutes(["Mesa-libGL-devel", "glproto-devel"], 
                           ["Mesa-libGL-devel", "xorgproto-devel"], update=True, check=True)
```

I checked using a `opensuse/tumbleweed` docker image and looks like it works fine, maybe you could check locally with the opengl recipe and report us back in case I'm missing something?

 I have opened a PR to the opengl/system recipe with these changes in Conan Center Index but I'll wait for your feedback: https://github.com/conan-io/conan-center-index/pull/23015

**Comment by neobrain** [comment]
@czoido For `opengl/system` specifically, this seems to work fine! I can install that package without overriding any `tools.system.package_manager` settings now.

The log output is a bit weird, but that's a very minor issue in comparison:
```
Mesa-libGL-devel-23.3.6-367.1.x86_64
package glproto-devel is not installed
Root privileges are required to run this command.
Mesa-libGL-devel-23.3.6-367.1.x86_64
xorgproto-devel-2023.2-1.2.noarch
opengl/system: System requirements:  already installed
```

I'm not sure this is the most elegant solution (particularly since it's not future-proof regarding further package renames elsewhere), but I can also see the concern about unintended effects of the `--whatprovides` proposal.

**Comment by czoido** [comment]
> @czoido For `opengl/system` specifically, this seems to work fine! I can install that package without overriding any `tools.system.package_manager` settings now.
> 
> The log output is a bit weird, but that's a very minor issue in comparison:
> 
> ```
> Mesa-libGL-devel-23.3.6-367.1.x86_64
> package glproto-devel is not installed
> Root privileges are required to run this command.
> Mesa-libGL-devel-23.3.6-367.1.x86_64
> xorgproto-devel-2023.2-1.2.noarch
> opengl/system: System requirements:  already installed
> ```
> 
> I'm not sure this is the most elegant solution (particularly since it's not future-proof regarding further package renames elsewhere), but I can also see the concern about unintended effects of the `--whatprovides` proposal.

I'm glad that it works, we will merge the changes then. If there's any more problems with other package renames, please just ping us or make a PR to Conan Center Index and that should be fixed quickly.

**Comment by memsharded** [comment]
Ok, fantastic.
Then I think we can close this one, as it is solvable so far with the ``substitutes``, if there is need to reconsider, we can re-open or create a new ticket here to discuss the feature, or in ConanCenter to fix recipes.

Thanks very much all!
